week3/main.cpp

```diff
@@ -0,0 +1,130 @@
+        }
+        if (x != 0 && y != 0) { // North-West
+            WordsFromPrefix(index - 1 - NumberOfColumns, prefix + Grid[index - 1 - NumberOfColumns],
+                            NumberOfColumns, NumberOfRows, Grid, TheDictionary, Answer, SamePaths);
+        }
+        if (x != 0 && y != NumberOfRows - 1) { // South-West
+            WordsFromPrefix(index - 1 + NumberOfColumns, prefix + Grid[index - 1 + NumberOfColumns],
+                            NumberOfColumns, NumberOfRows, Grid, TheDictionary, Answer, SamePaths);
+        }
+        if (x != NumberOfColumns - 1) { // East
+            WordsFromPrefix(index + 1, prefix + Grid[index + 1], NumberOfColumns, NumberOfRows, Grid,
+                            TheDictionary, Answer, SamePaths);
+        }
+        if (x != NumberOfColumns - 1 && y != 0) { // North-East
+            WordsFromPrefix(index + 1 - NumberOfColumns, prefix + Grid[index + 1 - NumberOfColumns], NumberOfColumns,
+                            NumberOfRows, Grid, TheDictionary, Answer, SamePaths);
+        }
+        if (x != NumberOfColumns - 1 && y != NumberOfRows - 1) { // South-East
+            WordsFromPrefix(index + 1 + NumberOfColumns, prefix + Grid[index + 1 + NumberOfColumns], NumberOfColumns,
+                            NumberOfRows, Grid, TheDictionary, Answer, SamePaths);
+        }
+        if (y != 0) { //North
+            WordsFromPrefix(index - NumberOfColumns, prefix + Grid[index - NumberOfColumns],
+                            NumberOfColumns, NumberOfRows, Grid, TheDictionary, Answer, SamePaths);
+        }
+        if (y != NumberOfRows - 1) { // South
+            WordsFromPrefix(index + NumberOfColumns, prefix + Grid[index + NumberOfColumns],
+                            NumberOfColumns, NumberOfRows, Grid, TheDictionary, Answer, SamePaths);
+        }
+    }
+}
+
+
+std::set<std::string> WordsFound(int NumberOfRows, int NumberOfColumns, char* Grid,
+                                 Dictionary &TheDictionary) {
+    // launch search from every cell
+    std::set<std::string> Answer;
+    std::string word1 = "";
+    std::set<std::pair<int, std::string>> SamePaths;
+    for(int i = 0; i < NumberOfColumns*NumberOfRows; ++i) {
+        word1 = Grid[i];
+        WordsFromPrefix(i, word1, NumberOfColumns, NumberOfRows, Grid, TheDictionary, Answer, SamePaths);
+    }
+    return Answer;
+}
+
+
+int main() {
+    // input data
+    std::cout << "Let's make a grid" << std::endl;
+    std::cout << "Enter the number of rows: ";
+    int NumberOfRows;
+    std::cin >> NumberOfRows;
+    std::cout << std::endl << "Enter the number of columns: ";
+    int NumberOfColumns;
+    std::cin >> NumberOfColumns;
+    std::cout << std::endl;
+    std::cout << "Now enter the grid" << std::endl;
+    char* Grid = new char[NumberOfRows * NumberOfColumns];
+    for(int i = 0; i < NumberOfRows; ++i) {
+        for (int j = 0; j < NumberOfColumns; ++j) {
+            std::cin >> Grid[i*NumberOfColumns+j];
```

Comment:
More spaces around the operations? :)

`i*NumberOfColumns+j   -->   I * NumberOfColumns + j`

It's difficult to read: i and variable name with * between them merge into one strange word.